File: src/main/scala/nl.knaw.dans.easy.dd2d/DatasetEditor.scala
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.easy.dd2d

import nl.knaw.dans.easy.dd2d.mapping.{ AccessRights, License }
import nl.knaw.dans.ingest.core.legacy.MetadataObjectMapper
import nl.knaw.dans.lib.dataverse.DataverseClient
import nl.knaw.dans.lib.dataverse.model.dataset.Embargo
import nl.knaw.dans.lib.dataverse.model.file.FileMeta
import nl.knaw.dans.lib.error.{ TraversableTryExtensions, TryExtensions }
import nl.knaw.dans.lib.logging.DebugEnhancedLogging
import org.json4s.native.Serialization
import org.json4s.{ DefaultFormats, Formats }

import java.net.URI
import java.nio.file.{ Path, Paths }
import java.util.regex.Pattern
import java.util.{ Date, Optional }
import scala.util.{ Failure, Success, Try }

/**
 * Object that edits a dataset, a new draft.
 */
abstract class DatasetEditor(dataverseClient: DataverseClient, optFileExclusionPattern: Option[Pattern], zipFileHandler: ZipFileHandler) extends DebugEnhancedLogging {
  type PersistentId = String
  type DatasetId = Int
  implicit val jsonFormats: Formats = DefaultFormats

  /**
   * Performs the task.
   *
   * @return the persistentId of the dataset created or modified
   */
  def performEdit(): Try[PersistentId]

  protected def addFiles(persistentId: String, files: List[FileInfo]): Try[Map[Int, FileInfo]] = Try {
    trace(persistentId, files)
    files.map { f =>
      debug(s"Adding file, directoryLabel = ${ f.metadata.getDirectoryLabel }, label = ${ f.metadata.getLabel }")
      val id = addFile(persistentId, f).unsafeGetOrThrow
      dataverseClient.dataset(persistentId).awaitUnlock()
      id -> f
    }.toMap
  }

  private def addFile(doi: String, fileInfo: FileInfo): Try[Int] = {
    val result = for {
      r <- {
        debug(s"Uploading file: $fileInfo")
        val optWrappedZip = zipFileHandler.wrapIfZipFile(fileInfo.file)
        val r = Try(dataverseClient.dataset(doi).addFileItem(
          Optional.of(optWrappedZip.getOrElse(fileInfo.file).toJava), // TODO what about closing files?
          Optional.of(Serialization.write(fileInfo.metadata))
        ))
        optWrappedZip.foreach(_.delete(swallowIOExceptions = true))
        r
      }
      files <- Try(r.getData)
      triedId = Try(files.getFiles.get(0).getDataFile.getId)
      _ <- Try(dataverseClient.dataset(doi).awaitUnlock())
    } yield triedId
    debug(s"Result = $result")
    result.map(_.getOrElse(throw new IllegalStateException("Could not get DataFile ID from response")))
  }

  protected def getPathToFileInfo(deposit: Deposit): Try[Map[Path, FileInfo]] = {
    for {
      bagPathToFileInfo <- deposit.getPathToFileInfo
      pathToFileInfo = bagPathToFileInfo.map { case (bagPath, fileInfo) => Paths.get("data").relativize(bagPath) -> fileInfo }
      filteredPathToFileInfo = excludeFiles(pathToFileInfo)
    } yield filteredPathToFileInfo
  }

  private def excludeFiles(p2fi: Map[Path, FileInfo]): Map[Path, FileInfo] = {
    trace(p2fi)
    p2fi.toList.filter {
      case (p, _) =>
        val foundMatch = optFileExclusionPattern.forall(_.matcher(p.toString).matches())
        if (foundMatch) logger.info(s"Excluding file: ${ p.toString }")
        !foundMatch
    }.toMap
  }

  protected def updateFileMetadata(databaseIdToFileInfo: Map[Int, FileMeta]): Try[Unit] = {
    trace(databaseIdToFileInfo)
    databaseIdToFileInfo.map { case (id, fileMeta) =>
      val json = Serialization.write(fileMeta)
      debug(s"id = $id, json = $json")
      val r = Try(dataverseClient.file(id).updateMetadata(json))
      debug(s"id = $id, result = $r")
      r
    }.collectResults.map(_ => ())
  }

  protected def configureEnableAccessRequests(deposit: Deposit, persistendId: PersistentId, canEnable: Boolean): Try[Unit] = {
    for {
      ddm <- deposit.tryDdm
      files <- deposit.tryFilesXml
      enable = AccessRights.isEnableRequests((ddm \ "profile" \ "accessRights").head, files)
      _ = logger.trace("AccessRequests enable "+ enable + " can " +canEnable)
      _ <- if (!enable) Try(dataverseClient.accessRequests(persistendId).disable())
           else if (canEnable) Try(dataverseClient.accessRequests(persistendId).enable())
                else Success(())
    } yield ()
  }

  protected def licenseAsJson(supportedLicenses: List[URI])(variantToNormalized: Map[String, String])(deposit: Deposit): Try[String] = {
    trace(deposit)
    for {
      ddm <- deposit.tryDdm
      license <- (ddm \ "dcmiMetadata" \ "license").find(License.isLicenseUri).map(Success(_)).getOrElse(Failure(RejectedDepositException(deposit, "No license specified")))
      uri = License.getLicenseUri(supportedLicenses)(variantToNormalized)(license).toASCIIString
    } yield s"""
                |{ "http://schema.org/license": "$uri" }
                |""".stripMargin
  }

  protected def isEmbargo(date: Date): Boolean = {
    date.compareTo(new Date()) > 0
  }

  protected def embargoFiles(persistendId: PersistentId, dateAvailable: Date, fileIds: List[Int]): Try[Unit] = {
    trace(persistendId, fileIds)
    val embargo = new Embargo(dateAvailableFormat.format(dateAvailable), "", fileIds.toArray)
    Try(dataverseClient.dataset(persistendId).setEmbargo(embargo))
  }

  protected def deleteDraftIfExists(persistentId: String): Unit = {
    val result = for {
      v <- Try(dataverseClient.dataset(persistentId).getLatestVersion.getData)
      _ = logger.trace("deleting draft")
      _ <- if (v.getLatestVersion.getVersionState.contains("DRAFT"))
             deleteDraft(persistentId)
           else Success(())
    } yield ()
    result.doIfFailure {
      case e => logger.warn("Could not delete draft", e)
    }
  }

  private def deleteDraft(persistentId: PersistentId): Try[Unit] = {
    for {
      _ <- Try(dataverseClient.dataset(persistentId).deleteDraft())
      _ = logger.info(s"DRAFT deleted")
    } yield ()
  }
}

File: src/main/java/nl/knaw/dans/ingest/core/service/DatasetEditor.java
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.ingest.core.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import nl.knaw.dans.ingest.core.dataverse.DatasetService;
import nl.knaw.dans.ingest.core.domain.Deposit;
import nl.knaw.dans.ingest.core.domain.FileInfo;
import nl.knaw.dans.ingest.core.exception.RejectedDepositException;
import nl.knaw.dans.ingest.core.service.mapper.mapping.FileElement;
import nl.knaw.dans.ingest.core.service.mapper.mapping.License;
import nl.knaw.dans.lib.dataverse.DataverseClient;
import nl.knaw.dans.lib.dataverse.DataverseException;
import nl.knaw.dans.lib.dataverse.Version;
import nl.knaw.dans.lib.dataverse.model.dataset.Dataset;
import nl.knaw.dans.lib.dataverse.model.file.DataFile;
import nl.knaw.dans.lib.dataverse.model.file.FileMeta;
import org.apache.commons.codec.digest.DigestUtils;
import org.w3c.dom.Node;

import java.io.FileInputStream;
import java.io.IOException;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeParseException;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Slf4j
public abstract class DatasetEditor {

    public static final String ORIGINAL_METADATA_ZIP = "original-metadata.zip";
    protected static final List<String> embargoExclusions = Arrays.asList("easy-migration.zip", ORIGINAL_METADATA_ZIP);

    protected final DataverseClient dataverseClient;
    protected final boolean isMigration;
    protected final Dataset dataset;
    protected final Deposit deposit;
    protected final List<URI> supportedLicenses;

    protected final Pattern fileExclusionPattern;
    protected final ZipFileHandler zipFileHandler;

    protected final ObjectMapper objectMapper;
    protected final DatasetService datasetService;

    protected final String vaultMetadataKey;

    protected DatasetEditor(boolean isMigration,
        Dataset dataset,
        Deposit deposit,
        List<URI> supportedLicenses,
        Pattern fileExclusionPattern,
        ZipFileHandler zipFileHandler,
        ObjectMapper objectMapper,
        DatasetService datasetService,
        String vaultMetadataKey) {
        this.dataverseClient = datasetService._getClient();
        this.isMigration = isMigration;
        this.dataset = dataset;
        this.deposit = deposit;
        this.supportedLicenses = supportedLicenses;
        this.fileExclusionPattern = fileExclusionPattern;
        this.zipFileHandler = zipFileHandler;
        this.objectMapper = objectMapper;
        this.datasetService = datasetService;
        this.vaultMetadataKey = vaultMetadataKey;
    }

    private static Instant parseDate(String value) {
        try {
            log.debug("Trying to parse {} as LocalDate", value);
            return LocalDate.parse(value).atStartOfDay(ZoneId.systemDefault()).toInstant();
        }
        catch (DateTimeParseException e) {
            try {
                log.debug("Trying to parse {} as ZonedDateTime", value);
                return ZonedDateTime.parse(value).toInstant();
            }
            catch (DateTimeParseException ee) {
                log.debug("Trying to parse {} as LocalDateTime", value);
                var id = ZoneId.systemDefault().getRules().getOffset(Instant.now());
                return LocalDateTime.parse(value).toInstant(id);
            }
        }
    }

    public abstract String performEdit() throws IOException, DataverseException, InterruptedException;

    Map<Integer, FileInfo> addFiles(String persistentId, Collection<FileInfo> fileInfos) throws IOException, DataverseException {
        var result = new HashMap<Integer, FileInfo>(fileInfos.size());

        for (var fileInfo : fileInfos) {
            log.debug("Adding file, directoryLabel = {}, label = {}",
                fileInfo.getMetadata().getDirectoryLabel(), fileInfo.getMetadata().getLabel());

            var id = addFile(persistentId, fileInfo);
            result.put(id, fileInfo);
        }

        return result;
    }

    protected FileInfo createOriginalMetadataFileInfo() throws IOException {
        var path = zipFileHandler.zipOriginalMetadata(deposit.getDdmPath(), deposit.getFilesXmlPath());
        var checksum = DigestUtils.sha1Hex(new FileInputStream(path.toFile()));
        var fileMeta = new FileMeta();
        fileMeta.setLabel(ORIGINAL_METADATA_ZIP);
        return new FileInfo(path, path, checksum, fileMeta);
    }

    private Integer addFile(String persistentId, FileInfo fileInfo) throws IOException, DataverseException {
        var dataset = dataverseClient.dataset(persistentId);
        var wrappedZip = zipFileHandler.wrapIfZipFile(fileInfo.getPhysicalPath());

        var file = wrappedZip.orElse(fileInfo.getPhysicalPath());
        if (log.isDebugEnabled()) {
            var metadata = objectMapper.writeValueAsString(fileInfo.getMetadata());
            log.debug("Adding file {} with metadata {}", file, metadata);
        }
        var result = dataset.addFile(file, fileInfo.getMetadata());
        log.debug("Called addFile for {}; result: {}", file, result);

        if (wrappedZip.isPresent()) {
            try {
                Files.deleteIfExists(wrappedZip.get());
            }
            catch (IOException e) {
                log.error("Unable to delete zipfile {}", wrappedZip.get(), e);
            }
        }

        log.debug("Result: {}", result);
        return result.getData().getFiles().get(0).getDataFile().getId();
    }

    protected String getLicense(Node node) {
        return XPathEvaluator.nodes(node, "/ddm:DDM/ddm:dcmiMetadata/dcterms:license")
            .filter(License::isLicenseUri)
            .findFirst()
            .map(n -> License.getLicenseUri(supportedLicenses, n))
            .map(URI::toASCIIString)
            .orElseThrow(() -> new RejectedDepositException(deposit, "no license specified"));
    }

    protected String toJson(Map<String, String> input) throws JsonProcessingException {
        return objectMapper.writeValueAsString(input);
    }

    Map<Path, FileInfo> getFileInfo() {
        var files = FileElement.pathToFileInfo(deposit, isMigration);

        return files.entrySet().stream()
            .map(entry -> {
                // relativize the path
                var bagPath = entry.getKey();
                var fileInfo = entry.getValue();
                var newKey = Path.of("data").relativize(bagPath);

                return Map.entry(newKey, fileInfo);
            })
            .filter(entry -> {
                // remove entries that match the file exclusion pattern
                var path = entry.getKey().toString();

                return (fileExclusionPattern == null || !fileExclusionPattern.matcher(path).matches());
            })
            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    // FIL008, FIL009
    void embargoFiles(String persistentId, Instant dateAvailable) throws IOException, DataverseException {
        var now = Instant.now();

        if (!dateAvailable.isAfter(now)) {
            log.debug("Date available in the past, no embargo: {}", dateAvailable);
        }
        else {
            var api = dataverseClient.dataset(persistentId);
            var files = api.getFiles(Version.LATEST.toString()).getData();

            var items = files.stream()
                .filter(f -> !embargoExclusions.contains(f.getLabel()))
                .map(FileMeta::getDataFile)
                .map(DataFile::getId)
                .collect(Collectors.toList());

            embargoFiles(persistentId, dateAvailable, items);
        }
    }

    void embargoFiles(String persistentId, Instant dateAvailable, Collection<Integer> fileIds) throws IOException, DataverseException {
        var now = Instant.now();

        if (!dateAvailable.isAfter(now)) {
            log.debug("Date available in the past, no embargo: {}", dateAvailable);
        }
        else if (fileIds.size() == 0) {
            log.debug("No files to embargo");
        }
        else {
            datasetService.setEmbargo(persistentId, dateAvailable, fileIds);
        }
    }

    Instant getDateAvailable(Deposit deposit) {
        return XPathEvaluator.strings(deposit.getDdm(), "/ddm:DDM/ddm:profile/ddm:available")
            .map(DatasetEditor::parseDate)
            .findFirst()
            .orElseThrow(() -> new IllegalArgumentException("Deposit without a ddm:available element"));
    }
}
